datafusion/core/src/physical_plan/repartition/distributor_channels.rs

```diff
@@ -0,0 +1,669 @@
+
+        // both channels non-empty => gate closed
+
+        let mut send_fut = txs[1].send("1_b");
+        let waker = poll_pending(&mut send_fut);
+
+        // drain channel 0
+        assert_eq!(poll_ready(&mut rxs[0].recv()), Some("0_a"),);
+        poll_pending(&mut send_fut);
+        assert_eq!(poll_ready(&mut rxs[0].recv()), Some("0_b"),);
+
+        // channel 0 empty => gate open
+        assert!(waker.woken());
+        poll_ready(&mut send_fut).unwrap();
+    }
+
+    #[test]
+    fn test_close_channel_by_dropping_tx() {
+        let (mut txs, mut rxs) = channels(2);
+
+        let tx0 = txs.remove(0);
+        let tx1 = txs.remove(0);
+        let tx0_clone = tx0.clone();
+
+        let mut recv_fut = rxs[0].recv();
+
+        poll_ready(&mut tx1.send("a")).unwrap();
+        let recv_waker = poll_pending(&mut recv_fut);
+
+        // drop original sender
+        drop(tx0);
+
+        // not yet closed (there's a clone left)
+        assert!(!recv_waker.woken());
+        poll_ready(&mut tx1.send("b")).unwrap();
+        let recv_waker = poll_pending(&mut recv_fut);
+
+        // create new clone
+        let tx0_clone2 = tx0_clone.clone();
+        assert!(!recv_waker.woken());
+        poll_ready(&mut tx1.send("c")).unwrap();
+        let recv_waker = poll_pending(&mut recv_fut);
+
+        // drop first clone
+        drop(tx0_clone);
+        assert!(!recv_waker.woken());
+        poll_ready(&mut tx1.send("d")).unwrap();
+        let recv_waker = poll_pending(&mut recv_fut);
+
+        // drop last clone
+        drop(tx0_clone2);
+
+        // channel closed => also close gate
+        poll_pending(&mut tx1.send("e"));
+        assert!(recv_waker.woken());
+        assert_eq!(poll_ready(&mut recv_fut), None,);
+    }
+
+    #[test]
+    fn test_close_channel_by_dropping_rx_on_open_gate() {
+        let (txs, mut rxs) = channels(2);
+
+        let rx0 = rxs.remove(0);
+        let _rx1 = rxs.remove(0);
+
+        poll_ready(&mut txs[1].send("a")).unwrap();
+
+        // drop receiver => also close gate
+        drop(rx0);
+
+        poll_pending(&mut txs[1].send("b"));
+        assert_eq!(poll_ready(&mut txs[0].send("foo")), Err(SendError("foo")),);
+    }
+
+    #[test]
+    fn test_close_channel_by_dropping_rx_on_closed_gate() {
+        let (txs, mut rxs) = channels(2);
+
+        let rx0 = rxs.remove(0);
+        let _rx1 = rxs.remove(0);
+
+        // fill both channels
+        poll_ready(&mut txs[0].send("0_a")).unwrap();
+        poll_ready(&mut txs[1].send("1_a")).unwrap();
+
+        let mut send_fut0 = txs[0].send("0_b");
+        let mut send_fut1 = txs[1].send("1_b");
+        let waker0 = poll_pending(&mut send_fut0);
+        let waker1 = poll_pending(&mut send_fut1);
+
+        // drop receiver
+        drop(rx0);
+
+        assert!(waker0.woken());
+        assert!(!waker1.woken());
+        assert_eq!(poll_ready(&mut send_fut0), Err(SendError("0_b")),);
+        poll_pending(&mut send_fut1);
```

Comment:
I don't really see what this gives us (I'm not a massive fan of testing every single interaction in every single test), but since it's a tiny addition, I've added it anyways.

Reply:
I was thinking it would provide coverage that even after one channel is closed then other non closed channels can continue to receive previously buffered things and don't get shut down early)